### ISSUE TITLE
Improve folder structure

### ISSUE THREAD

**Issue by Stijnkool** [inmanta/web-console#2022]
## TL;DR
We will introduce a slices folder. The 3 layer structure (core-data-ui) will also be reflected in each slice. This will allow us to put related files together into a slice.

### Todo

We will gradually update the codebase to this new structure. The root folders we currently have are still correct.
We can just create new slices as we need them. When you touch component in the old structure, you can create a slice for it. The pages are obvious candidates. But more components might require their own slice.

- [ ] Update documentation

## Dependency structure
For imports (dependency hierarchy) there should be 3 layers (like an onion):
- Core (domain, contracts, query, command, ...) -> depends on nothing
- Data (data transformations and implementations of contracts) -> depends on Core
- UI (pages, components) -> depends on Data & Core

Dependencies go in 1 direction: `UI -> Data -> Core`.
Core can be used by all. UI can't be used by anyone as it is the end result.
By this definition, the Routing module should be placed in the UI folder as it uses Core, but defines UI and is used by the Root UI.

## Problem
- Adding a page requires changes in a lot of files, in a lot of different folders. This is not ideal.
- Changing something is harder because related files are not located together.
  - the domain type is defined somewhere in Core/Domain 
  - the query is defined in Core/Query
  - The data manager is defined in Data/Managers
  - The UI is defined in UI
  - These are all related but not located close to eachother.

This is because we use a high level folder structure by layer. The layers are good. And still useful.
But we also want to put related files together.

## Solution
We will introduce a Slices folder. This contains a folder per slice. A slice usually has some UI and a data requirement.
It can also have some other logic and domain type definitions, but that is not required.

### Structure
- Core (common)
- Data (common)
- UI (common, contains Root)
- Routing (* or inside UI to be consistent with our 3 layers)
- Slices (repeat the 3 layers for each slice)
  - Catalog
    - Core (slice specific)
      - Query & Command
      - any types or contracts
    - Data (slice specific)
      - Managers
      - transformations
      - implementations
      - store slice definition
    - UI (slice specific)
      - The components

If a slice does not need a layer folder, just omit it. If something is used by 2 slices, move it to the respective common folder.
We can also put multiple components or pages in 1 slice. But that all depends on the use case.
Every layer within a slice does not need it own folder. It can also just be a file. (Core.ts, Data.ts)

## Routing
We might have to keep the PageManager out of the Routing module to avoid circular dependencies.
The page manager imports all other pages. But all views can have a dependency on Routing utilities.
If both are contained in the Routing module and within the Routing namespace, it might cause problems.
We can have Routing utilities separate. Or we can move the PageManager to the common UI folder.

## Test files
- Test files should have the `{identifier}.test.ts{?x}` format.
- Test functions should be defined with the `test()` api, not the `it()` api.
- Test files can be grouped in a `Test` folder, not the `Spec` folder.

We currently have 1 folder with test data.
We can now define test data next to the test files or domain files.
But once we have common data, it should still go in a common folder.
Imports are not really in issue for test data as this code is never included in the app itself.

## Imports
Moving the queries, commands and managers inside a slice folder puts related code more together.
But there are no direct dependencies between these because the dependencies are handled mostly by context.
It is still useful to have them together even though there are no direct imports between them.
Changing a query will probably also require a change in the UI and manager.

**Comment by wouterdb** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/core-5a7c152b04a1652024289b7b/issues/inmanta/web-console/2022?planning-poker) with ZenHub @bartv @Stijnkool

**Comment by wouterdb** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/core-5a7c152b04a1652024289b7b/issues/inmanta/web-console/2022?planning-poker) with ZenHub @andraskvr @bartv @Stijnkool

**Comment by Stijnkool** [comment]
@andraskvr Do I need to mention anything else? Do you agree with my write up?

**Comment by Stijnkool** [comment]
@bartv Doing this refactor will make merging with iso4 almost impossible. Or at least very hard.
Is it okay to only do bug fixes and security fixes on iso4? We will probably have to do them manually for iso4.

**Comment by andraskvr** [comment]
> @andraskvr Do I need to mention anything else? Do you agree with my write up?

It looks good :+1:
